db/migrate/20231210061330_add_created_at_to_fts_targets_with_index.rb

```diff
@@ -0,0 +1,106 @@
+# For auto load
+FullTextSearch::Migration
+
+class AddCreatedAtToFtsTargetsWithIndex < ActiveRecord::Migration[5.2]
+  def up
+    return if !table_exists?(:fts_targets)
+
+    add_column :fts_targets, :created_at, :timestamp
+
+    ActiveRecord::Base.transaction do
+      update_created_at_using_last_modified_at
+      update_created_at_using_created_on
+    end
+
+    if Redmine::Database.mysql?
+      add_index :fts_targets, :created_at
+    else
+      remove_index :fts_targets, name: "fts_targets_index_pgroonga"
+      add_index :fts_targets,
+                [:id,
+                 :source_id,
+                 :source_type_id,
+                 :project_id,
+                 :container_id,
+                 :container_type_id,
+                 :custom_field_id,
+                 :is_private,
+                 :last_modified_at,
+                 :created_at,
+                 :title,
+                 :content,
+                 :tag_ids],
+                using: "PGroonga",
+                with: "normalizer = 'NormalizerNFKC121'",
+                name: "fts_targets_index_pgroonga"
+    end
+  end
+
+  def down
+    return if !table_exists?(:fts_targets)
+
+    if Redmine::Database.mysql?
+      remove_index :fts_targets, :created_at
+    else
+      remove_index :fts_targets, name: "fts_targets_index_pgroonga"
+      add_index :fts_targets,
+                [:id,
+                 :source_id,
+                 :source_type_id,
+                 :project_id,
+                 :container_id,
+                 :container_type_id,
+                 :custom_field_id,
+                 :is_private,
+                 :last_modified_at,
+                 :title,
+                 :content,
+                 :tag_ids],
+                using: "PGroonga",
+                with: "normalizer = 'NormalizerNFKC121'",
+                name: "fts_targets_index_pgroonga"
+    end
+
+    remove_column :fts_targets, :created_at
+  end
+
+  private
+
+  class TmpTarget < ActiveRecord::Base
+    self.table_name = 'fts_targets'
+  end
+
+  class TmpType < ActiveRecord::Base
+    self.table_name = 'fts_types'
+  end
+
+  def update_created_at_using_last_modified_at
+    types = ['Attachment', 'Change', 'Changeset', 'CustomValue', 'Document', 'Journal', 'News']
+
+    TmpType.where(name: types).pluck(:id).each do |type_id|
+      TmpTarget.where(source_type_id: type_id).in_batches do |targets|
+        targets.update_all('created_at = last_modified_at')
+      end
+    end
+  end
+
+  def update_created_at_using_created_on
+    types = ['Issue', 'Message', 'Project', 'WikiPage']
+
```

Comment:
Does this work?

```ruby
class Issue < ActiveRecord::Base
end

class FtsTarget < ActiveRecord::Base
end

class FtsType < ActiveRecord::Base
end

FtsTarget.
  where(source_type_id: FtsType.where(name: "Issue").select(:id)).
  update_all(created_at: Issue.where(id: FtsTarget.arel_table[:source_id]).select(:created_on))
```

Reply:
fix: [00926ae](https://github.com/clear-code/redmine_full_text_search/pull/118/commits/00926ae95772cb34e1f4a59f22d91bfb49ebcb89)
Despite attempting your suggestion, it unfortunately did not work. So, I switched to generating the subquery using the Arel interface. This approach is functional, although there's an issue with embedding the subquery into the SQL as a variable.
- I think this subquery is generated by the Arel interface so it will be more safe than the previous implementation.
```console
> FtsTarget.where(source_type_id: FtsType.where(name: "Issue").select(:id)).update_all(created_at: Issue.where(id: FtsTarget.arel_table[:source_id]).select(:created_on))
UPDATE "fts_targets" SET "created_at" = $1 WHERE "fts_targets"."source_type_id" IN (SELECT "fts_types"."id" FROM "fts_types" WHERE "fts_types"."name" = $2)  [["created_at", nil], ["name", "Issue"]]
```